packages/liveblocks-core/src/connection.ts

```diff
@@ -0,0 +1,756 @@
+      didConnect.notify();
+    }
+    oldPublicStatus = newPublicStatus;
+  });
+
+  return {
+    statusDidChange: statusDidChange.observable,
+    didConnect: didConnect.observable,
+    didDisconnect: didDisconnect.observable,
+  };
+}
+
+function createStateMachine<T extends BaseAuthResult>(delegates: Delegates<T>) {
+  // Create observable event sources, which this machine will call into when
+  // specific events happen
+  const onMessage = makeEventSource<IWebSocketMessageEvent>();
+
+  // Emitted whenever the server deliberately closes the connection for
+  // a specific Liveblocks reason
+  const onLiveblocksError = makeEventSource<LiveblocksError>();
+
+  const initialContext: Context & { token: T | null } = {
+    token: null,
+    socket: null,
+
+    // Bumped to the next "tier" every time a connection attempt fails (no matter
+    // whether this is for the authentication server or the websocket server).
+    // Reset every time a connection succeeded.
+    backoffDelay: LOW_DELAY,
+    // numRetries: 0,
+  };
+
+  const fsm = new FSM<Context, Event, State>(initialContext)
+    .addState("@idle.initial")
+    .addState("@idle.failed")
+    .addState("@auth.busy")
+    .addState("@auth.backoff")
+    .addState("@connecting.busy")
+    .addState("@connecting.backoff")
+    .addState("@ok.connected")
+    .addState("@ok.awaiting-pong");
+
+  //
+  // Configure events that can happen from anywhere
+  //
+  // It's always possible to explicitly get a .reconnect() or .disconnect()
+  // from the user.
+  //
+  fsm.addTransitions("*", {
+    RECONNECT: {
+      target: "@auth.backoff",
+      assign: increaseBackoffDelay,
+    },
+
+    DISCONNECT: "@idle.initial",
+  });
+
+  //
+  // Configure the @idle.* states
+  //
+  fsm.addTransitions("@idle.*", {
+    CONNECT: (_, ctx) =>
+      // If we still have a known token, try to reconnect to the socket directly,
+      // otherwise, try to obtain a new token
+      ctx.token !== null ? "@connecting.busy" : "@auth.busy",
+  });
+
+  //
+  // Configure the @auth.* states
+  //
+  fsm
+    .addTransitions("@auth.backoff", {
+      NAVIGATOR_ONLINE: {
+        target: "@auth.busy",
+        assign: { backoffDelay: LOW_DELAY },
+      },
+    })
+    .addTimedTransition(
+      "@auth.backoff",
+      (ctx) => ctx.backoffDelay,
+      "@auth.busy"
+    )
+
+    .onEnterAsync(
+      "@auth.busy",
+
+      () =>
+        Promise.race([delegates.authenticate(), timeoutAfter(AUTH_TIMEOUT)]),
+
+      // On successful authentication
+      (okEvent) => ({
+        target: "@connecting.busy",
+        assign: {
+          token: okEvent.data as BaseAuthResult,
+          backoffDelay: LOW_DELAY,
+        },
+      }),
+
+      // Auth failed
+      (failedEvent) =>
+        failedEvent.reason instanceof UnauthorizedError
+          ? {
+              target: "@idle.failed",
+              effect: () =>
+                console.error(
+                  `Unauthorized, will stop retrying: ${
```

Comment:
I don't know that we have to explicitly say we'll stop retrying here.

Reply:
Fixed in c44a469f6f1045b615398780765a1c15ea9e8de1 👍